### PR TITLE
Add SpriteBatch API

### PR DESCRIPTION
**Pull request by erf**
This is an PR for #282.

- added a `SpriteBatch` class which buffers rects, transforms and optional colors, and renders using `Canvas.drawAtlas`
- added a `SpriteBatchComponent` which accepts a `SpriteBatch`
- added an Example which render a bunch of sprites in various ways ( using `SpriteBatch` )

EDIT: 
- added new doc entry `SpriteBatch` under `images.md`
- updated CHANGELOG

### CODE REVIEW

**Comment by erf** [comment]
Ok, i will look into those changes. 

Here is a screenshot:
Simulator Screen Shot - iPhone 11 Pro - 2020-04-16 at 14 32 01

**Comment by erf** [comment]
I added a doc entry and updated changelog.

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,9 @@
 
 ## [next]
 
+## 0.20.0
```
Comment (erickzanardo) [inline CHANGELOG.md:5]: Leave this on the `[next]` entry, which version will be is defined when we release the version.

File: doc/images.md [hunk 2]
```diff
@@ -45,6 +45,18 @@ Sprites can also be used as widgets, to do so, just use `Flame.util.spriteAsWidg
...
+
+If you have an image atlas ( an image with smaller images inside ), and would like to render it effectively - `SpriteBatch` does the job.
+
+Give it the filename of the image, and then add rectangles which describes various part of the image, in addition to transforms ( position, scale and rotation ) and optional colors.
+
+You render it with an `Canvas` and an optional `Paint`, `BlendMode` and `CullRect`.
+
+An `SpriteBatchComponent` is also available for your convenience.
+
+See example [here](examples/sprite_batch).
```
Comment (erickzanardo) [inline doc/images.md:58]: Does this link work? I think it should be `docs/example/sprite_batch` doesn't it?

File: lib/sprite_batch.dart [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+  static final defaultTransform = RSTransform(1, 0, 0, 0);
+  static const defaultColor = const Color(0x00000000); // transparent
+
+  SpriteBatch(String fileName) {
+    Flame.images.load(fileName).then((image) => atlas = image);
+  }
+
+  SpriteBatch.fromImage(this.atlas);
+
+  int get width => atlas.width;
```
Comment (erickzanardo) [inline lib/sprite_batch.dart:25]: Just to make sure, I would verify for null here

Example
```dart
  int get width => atlas?.width;
```

Same thing for `height` and `size`

File: doc/images.md (hunk 2 above)
Comment (erf) [reply doc/images.md:58]: I think you are right. I looked at `examples/animation_widget` which is also wrong then? Will update that also.

File: lib/sprite_batch.dart (hunk 3 above)
Comment (erf) [reply lib/sprite_batch.dart:25]: do `atlas?.height ?? 0` or `atlas?.height` and return `null`

for size do:
```
  Size get size => atlas == null ? null : Size(width.toDouble(), height.toDouble());
```
or return size with 0 values ?

btw: i think i would prefer to pass an image to the constructor, which is not null, so we dont have to do async load stuff and null checks there. assets could be loaded outside and prepared beforehand

**Comment by erf** [comment]
Should i amend and force push changes as i fix them here?

**Comment by erf** [comment]
Rebased and force pushed changes with fixes.

File: doc/examples/sprite_batch/lib/main.dart [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+      final y = r.nextInt(size.height ~/ 2).toDouble() + size.height / 2.0;
+      spriteBatch.add(
+        rect: Rect.fromLTWH(sx, sy, 128, 128),
+        offset: Offset(x - 64, y - 64),
+      );
+    }
+
+    add(SpriteBatchComponent.fromSpriteBatch(
+      spriteBatch,
+      //paint: Paint()..invertColors = true,
```
Comment (luanpotter) [inline doc/examples/sprite_batch/lib/main.dart:61]: please either uncomment or remove

File: doc/images.md [hunk 5]
```diff
@@ -43,7 +43,19 @@ All render methods from the Sprite class can receive a `Paint` instance on the o
 
 Sprites can also be used as widgets, to do so, just use `Flame.util.spriteAsWidget`
 
-A complete example of using sprite as widgets can be found [here](examples/animation_widget).
+A complete example of using sprite as widgets can be found [here](/doc/examples/animation_widget).
+
+## SpriteBatch
+
+If you have an image atlas ( an image with smaller images inside ), and would like to render it effectively - `SpriteBatch` does the job.
```
Comment (luanpotter) [inline doc/images.md:50]: 1. please don't add spaces inside parenthesis (like this).
2. we pretty much call "sprite atlas" spritesheets everywhere else. AFAICS this is the same thing just being rendered more efficiently. so maybe I'd change the text to be
> If you have a sprite sheet (also called an image atlas, which is ...), ...
just to make it more clear that spritesheet and atlas from the flame perspective are the same thing, this is just a new way to render it (correct me if I'm wrong)

File: doc/images.md [hunk 6]
```diff
@@ -43,7 +43,19 @@ All render methods from the Sprite class can receive a `Paint` instance on the o
...
+
+Give it the filename of the image, and then add rectangles which describes various part of the image, in addition to transforms ( position, scale and rotation ) and optional colors.
```
Comment (luanpotter) [inline doc/images.md:52]: (spaces in parenthesis)

File: lib/components/sprite_batch_component.dart [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+      canvas,
+      blendMode: blendMode,
+      cullRect: cullRect,
+      paint: paint,
+    );
+  }
+
+  @override
+  bool loaded() {
+    return spriteBatch != null && spriteBatch.atlas != null;
```
Comment (luanpotter) [inline lib/components/sprite_batch_component.dart:31]: you can do `spriteBatch?.atlas != null`. and then you can inline this function like this
```
bool loaded() => spriteBatch?.atlas != null;
```

File: doc/images.md (hunk 5 above)
Comment (erf) [reply doc/images.md:50]: Yeah, they are essentially doing the same thing, only SpriteSheet now has a bit more functionality. Should replace "SpriteSheet" with "SpriteBatch" later ( and perhaps rename SpriteBatch to SpriteSheet again )

File: lib/components/sprite_batch_component.dart (hunk 7 above)
Comment (erf) [reply lib/components/sprite_batch_component.dart:31]: could `loaded` be a `get` property here?

**Comment by erf** [comment]
pushed changes

**Comment by erf** [comment]
I added a new commit where i changed the constructor, given the image asset, and created an async factory method which creates and return a new `SpriteBatch` with a preloaded asses, as i commented in https://github.com/flame-engine/flame/issues/289#issuecomment-615374515 
In this way we always guarantee that the atlas image is loaded and we dont have to do null checks.

**Comment by erf** [comment]
Btw i wanted to pass optional named parameters for rects, transforms and colors to the spritebatch, but they become null, even if i declared them as empty lists in the class. Not sure why this is. I added a question on stackoverflow here:
https://stackoverflow.com/questions/61303278/optional-constructor-parameter-set-the-list-to-null-in-dart/61303489?noredirect=1#comment108449447_61303489

but i think it might be better to use named constructors at a later point, in order to create sprite sheets etc.

File: lib/sprite_batch.dart (hunk 3 above)
Comment (erickzanardo) [reply lib/sprite_batch.dart:25]: I think we should just return null, not 0 as default, just to avoid a NPE inside the method.

Comment (erf) [reply lib/sprite_batch.dart:25]: In the latest commit you dont have to check for null

File: lib/components/sprite_batch_component.dart [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+    spriteBatch.render(
+      canvas,
+      blendMode: blendMode,
+      cullRect: cullRect,
+      paint: paint,
+    );
+  }
+
+  @override
+  bool loaded() => true;
```
Comment (luanpotter) [inline lib/components/sprite_batch_component.dart:30]: this returns true by default, no need to override it

**Comment by erf** [comment]
Ok, should be good to go.

**Comment by renancaraujo** [comment]
This is amazing. Good to go!

**Pull request merged**